framework/common/encryption.cpp

```diff
@@ -223,61 +223,112 @@ bool hmac::digest(uint8_t *digest)
     return HMAC_Final(m_ctx, digest, &digest_length);
 }
 
-bool aes_encrypt(const uint8_t *key, const uint8_t *iv, uint8_t *data, uint32_t data_len)
+bool aes_encrypt(const uint8_t *key, const uint8_t *iv, uint8_t *plaintext, int plen,
+                 uint8_t *ciphertext, int &clen)
 {
     EVP_CIPHER_CTX *ctx;
+    int len;
 
-    int clen, len;
-    uint8_t buf[128];
-
+    /* Create and initialise the context */
     ctx = EVP_CIPHER_CTX_new();
     if (NULL == ctx) {
+        MAPF_ERR("Failed to create context");
         return false;
     }
+
+    /*
+     * Initialise the encryption operation.
+     * IMPORTANT - key and IV should be
+     * The same as block size, which 16 bytes since we are using 
+     * 128 bit AES (i.e. a 128 bit key). The IV size for *most* modes is
+     * the same as the block size.
+     */
     if (EVP_EncryptInit_ex(ctx, EVP_aes_128_cbc(), NULL, key, iv) != 1) {
+        MAPF_ERR("EVP_EncryptInit_ex Failed");
+        EVP_CIPHER_CTX_free(ctx);
         return false;
     }
-    EVP_CIPHER_CTX_set_padding(ctx, 0);
 
-    clen = data_len;
-    if (EVP_EncryptUpdate(ctx, data, &clen, data, data_len) != 1 || clen != (int)data_len) {
+    /*
+     * Provide the message to be encrypted, and obtain the encrypted output.
+     * EVP_EncryptUpdate() encrypts inl bytes from the buffer in and writes
+     * the encrypted version to out. The amount of data written depends on
+     * the block alignment of the encrypted data: as a result the amount of
+     * data written may be anything from zero bytes to
+     * (inl + cipher_block_size - 1) so out should contain sufficient room.
+     * In this case - clen has to be >= plaintext_len + 15)
+     */
+    if ((clen < plen + 15) || (EVP_EncryptUpdate(ctx, ciphertext, &len, plaintext, plen) != 1)) {
```

Comment:
The `clen < plen + 15` check is a check of the input parameters, so it doesn't make sense IMO to do it in the same condition as the EVP_EncryptUpdate call.

Also, the check is not semantically correct: there should just be sufficient space to do padding, not always 15 additional bytes. But I'm OK with using this approximation, just note it in the comment.

Reply:
I disagree - IMHO it is more appropriate here since it is our take on the `EVP_EncryptUpdate` function which *will* fail if we don't give it enough buffer space.

Reply:
Fixed.